staging/src/k8s.io/api/admissionregistration/v1alpha1/types.go

```diff
@@ -198,26 +198,25 @@ const (
 // AdmissionHookClientConfig contains the information to make a TLS
 // connection with the webhook
 type AdmissionHookClientConfig struct {
-	// Service is a reference to the service for this webhook. If there is only
-	// one port open for the service, that port will be used. If there are multiple
-	// ports open, port 443 will be used if it is open, otherwise it is an error.
-	// Required
-	Service ServiceReference `json:"service" protobuf:"bytes,1,opt,name=service"`
-
-	// URLPath is an optional field that specifies the URL path to use when posting the AdmissionReview object.
-	URLPath string `json:"urlPath" protobuf:"bytes,3,opt,name=urlPath"`
+	// `url` gives the location of the webhook, in standard URL form
+	// (`[scheme://]host:port/path`).
+	//
+	// If the host refers to a service running in the cluster, it must be
+	// of the form `servicename.namespace.svc.`. Do not include the
+	// "cluster.local" suffix, as that is a configurable detail and would
+	// render your config non-portable; apiservers will route to the
+	// service correctly.
+	//
+	// If the scheme is present, it must be "https://".
+	//
+	// A path is optional, and if present may be any string permissible in
+	// a URL. You may use the path to pass an arbitrary string to the
+	// webhook, for example, a cluster identifier.
+	//
+	// Required.
+	URL string
```

Comment:
I like this change.

Reply:
Unfortunately I'm hearing from @bowei that there is no safe standard DNS form that we could parse? Bowei, please say I misunderstood :( :(

Reply:
I'm unsure what is meant by that. I'd expect to fully specify the URL, including the scheme: `scheme://host[:port][/path]`. What is ambiguous about that?

Reply:
Do we need to distinguish between a service exported by the cluster vs a server with a DNS name reachable on the network? 

What Daniel is referring to is that we do not want to bake in the current DNS schema into the system in case the schema changes. This is a particularly subtle place to have to be maintained if any changes are done.

Any alternative is to have a different scheme for identifying cluster only services, something like: `k8s-service://`.

Reply:
@liggit, kube-apiserver must be able to tell whether the DNS name refers to something in the cluster or out of the cluster, because it can't assume the presence of DNS resolution for in-cluster things (or, currently, routes, though that will change in a few quarters).

Reply:
@cheftako proposed another solution, which is to use the external hostname field in a service object to solve the cert checking problem we have right now.

Reply:
@bowei I'm not sure if using a different scheme helps-- wouldn't kube-apiserver still have to parse the hostname to figure out what service it refers to?

Reply:
Sounds like people think using an invented TLD is basically a terrible idea:

https://serverfault.com/questions/17255/top-level-domain-domain-suffix-for-private-network

Reply:
We could use the .invalid TLD, hehe

https://en.wikipedia.org/wiki/.test

Reply:
**Option 1**: How about `<name>.<namespace>.svc.k8s.invalid` as a portable unambiguous service description?

Significant con: extension apiservers would also have to parse this and convert to the DNS system in use in the cluster. (mitigating factor: this not much worse than today, where they construct this from the namespace and name)

**Option 2**: use `<name>.<namespace>.svc.` as currently written in the comment.

Con: @bowei has to pick up the pieces when his team tries to fix cluster DNS naming. (mitigating factor: the ship has probably already sailed.) I'm totally fine with this but @bowei may not be.

**Option 3**: Keep the current system, and if the service actually refers to an external location, it must have the external hostname field set, and in that case all apiservers can use that DNS name directly. Otherwise kube-apiserver uses the current resolution mechanism, and extension apiservers can either do that or convert to in-cluster-DNS form (whatever that flavor du jour is) and use DNS.

Con: service reference is kinda ugly.

Reply:
Actually, the con I've listed under option 2 applies to everything, since no matter what we do, extension-apiservers will be constructing a DNS name, and this would have to be fixed if a change to this representation were made.

Reply:
Over lunch I thought of **Option 4**:

1. Any domain with a terminating `.` will be treated as an external domain (e.g. `googleapis.com.`)
2. Any domain without a terminating `.` **and** that ends in `.svc` will be treated as an in-cluster service reference.

Reply:
Er, actually I was going to say that any domain without a terminating `.` will be treated as a prefix for the cluster's own identifying domain, which is presumably a concept we have or will have at some point in the future.

Reply:
Suggestion:
- at most one of `Service ServiceReference` or `URL string` may be set.  
- if `URL string` is used, normal DNS resolution is used, then no magic happens, and  `<name>.<namespace>.svc.` is not expected to work.

Reply:
This lets the user chose which is the lesser evil: layering violation or non-portable configuration.

Reply:
I've pushed a commit that keeps both. I don't like it very much. Why not maintain the Service field, and make the external hostname field of that work correctly?

Reply:
@lavalamp @deads2k 

I'm in favor of only a service reference, and opposed to a URL. As we can see from this discussion, URLs obscure information that needs to be first-class. In particular, if something generates or munges the Service name, it needs to know how to populate this field. Everywhere else in the API we use object references or names. 

So, option 3.

Reply:
For self-hosted webhooks, I can live with service references, resolved as they are today. 

For non-cluster-hosted webhooks, it doesn't make sense to require the indirection of using a service reference pointing to a service which itself points to an external hostname. That still requires the external hostname to be resolvable, *and* requires all consumers of admission webhook config (even ones in pods that could have used DNS for in- and out- of cluster resolution) to have custom dialers and read permission on the service API objects. 

> @erictune
> Suggestion:
> at most one of Service ServiceReference or URL string may be set.
> if URL string is used, normal DNS resolution is used, then no magic happens, and `<name>.<namespace>.svc.` is not expected to work.
> This lets the user chose which is the lesser evil: layering violation or non-portable configuration.

That seems reasonable to me (and it's not a layer violation to use DNS to resolve off-cluster webhook hostnames)

Reply:
It seems like I never submitted my reply to eric's comment above, but Jordan and Eric's approach seems to be reasonable.

Reply:
What I care most about is that we keep service/object references for self-hosted webhooks.